File: wrap/src/main/java/com/github/ykiselev/wrap/Wrap.java
package com.github.ykiselev.wrap;

import static java.util.Objects.requireNonNull;

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
public abstract class Wrap<T> implements AutoCloseable {

    private final T value;

    public Wrap(T value) {
        this.value = requireNonNull(value);
    }

    public final T value() {
        return value;
    }

    @Override
    public abstract void close();
}

File: wrap/src/main/kotlin/com/github/ykiselev/wrap/Wrap.kt
package com.github.ykiselev.wrap

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
abstract class Wrap<T>(private val value: T) : AutoCloseable {

    fun value(): T =
        value

    abstract override fun close()

    override fun toString(): String =
        "Wrap{$value}"
}